### PR TITLE
Faml 758 include errors on direct entry 2nd attempt

### PR DESCRIPTION
**Pull request by ch-rfishwick**
Use cached data, set during the journey, to make certain pages inaccessible until a certain point has been reached. In this manner users will not be able to arbitrarily enter a point in the user journey without first reaching it in the required manner. 
If they complete or restart their journey then they will similarly be unable to directly enter an arbitrary point in it.

### CODE REVIEW

File: server/routes/report.js [hunk 1]
```diff
@@ -197,9 +242,12 @@ router.post('/report-a-discrepancy/company-number', (req, res) => {
         return pscDiscrepancyService.saveCompanyNumber(Utility.getOAuthToken(req), data);
       }
     }).then(_ => {
+      const pscCache = cacheService.getCachedDataFromSession(req.session);
       if (secureFlag === true) {
+        pscCache.passedCompanyNumberWithSecureFlag = true;
         return res.redirect(302, '/report-a-discrepancy/secure-psc');
       }
+      pscCache.passedCompanyNumber = true;
       return res.redirect(302, '/report-a-discrepancy/confirm-company');
```
Comment (CH-JonathanBlake) [inline server/routes/report.js:251]: Could we else this, as I think it would be more obvious from a scan that these are exclusive scenarios

**Review by ICraig-CH — COMMENTED** [review]
you need to add a title to the res.render(......) lines because otherwise it will be like the error page in terms of displaying the matomo events.

File: server/views/_partials/page_not_available.njk [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+{% block main_content %}
+    <div class="govuk-width-container ">
+            <main class="govuk-main-wrapper govuk-main-wrapper--auto-spacing" id="main-content" role="main">
+                <div class="govuk-grid-row">
+                    <div class="govuk-grid-column-two-thirds">
+                        <div class="text">
+                            <h1 class="govuk-heading-l">Page not available</h1>
+                        </div>
+
+                        <p>To access this page please <a class="govuk-link" href="/report-a-discrepancy#start">make a discrepancy report</a></p>
```
Comment (ICraig-CH) [inline server/views/_partials/page_not_available.njk:12]: Would you mind putting a Matomo event in for this link please?

**Comment by ch-rfishwick** [comment]
> you need to add a title to the res.render(......) lines because otherwise it will be like the error page in terms of displaying the matomo events.

how do you mean?

**Comment by ICraig-CH** [comment]
> > you need to add a title to the res.render(......) lines because otherwise it will be like the error page in terms of displaying the matomo events.
> 
> how do you mean?

like its done here :      res.render(`${routeViews}/company_number.njk`, { title: routeUtils.setPageTitle('What is the company number for the PSC with the discrepancy?') });
so that Matomo can pick up the title of the page you send it to

File: server/routes/report.js [hunk 3]
```diff
@@ -33,8 +33,25 @@ router.use((req, res, next) => {
...
+  pscCache.passedPscNameWithKnownKind = false;
+  pscCache.passedPscDiscrepancyTypes = false;
+  pscCache.passedDiscrepancyDetails = false;
+  pscCache.passedCheckYourAnswers = false;
+}
+
 router.get('(/report-a-discrepancy)?', (req, res, next) => {
   logger.info(`GET request to serve index page: ${req.path}`);
+  const pscCache = cacheService.getCachedDataFromSession(req.session);
+  clearPathTracking(pscCache);
```
Comment (CH-JonathanBlake) [inline server/routes/report.js:54]: We could move this section into a cacheService method.

Comment (ch-rfishwick) [reply server/routes/report.js:54]: You mean encapsulate the  `clearPathTracking()` method within `CacheService`?

Comment (ch-rfishwick) [reply server/routes/report.js]: Ok, I've extracted it, was that the sort of thing you meant?

File: server/routes/report.js (hunk 1 above)
Comment (ch-rfishwick) [reply server/routes/report.js]: changed

File: server/views/_partials/page_not_available.njk (hunk 2 above)
Comment (ch-rfishwick) [reply server/views/_partials/page_not_available.njk]: done

**Review by ICraig-CH — APPROVED** [review]

File: server/services/cache_service.js [hunk 4]
```diff
@@ -7,6 +7,21 @@ class CacheService {
...
+    this.getCachedDataFromSession(req.session).passedObligedEntityEmail = false;
+    this.getCachedDataFromSession(req.session).passedCompanyNumberWithSecureFlag = false;
+    this.getCachedDataFromSession(req.session).passedCompanyNumber = false;
+    this.getCachedDataFromSession(req.session).passedConfirmCompany = false;
+    this.getCachedDataFromSession(req.session).passedPscName = false;
+    this.getCachedDataFromSession(req.session).passedPscNameWithKnownKind = false;
+    this.getCachedDataFromSession(req.session).passedPscDiscrepancyTypes = false;
+    this.getCachedDataFromSession(req.session).passedDiscrepancyDetails = false;
+    this.getCachedDataFromSession(req.session).passedCheckYourAnswers = false;
+  }
```
Comment (CH-JonathanBlake) [inline server/services/cache_service.js:24]: even if you don't want to move this into the cache service, at least move the getCachedData query into a variable.

Comment (ch-rfishwick) [reply server/services/cache_service.js:24]: it is in the cache service

**Review by CH-JonathanBlake — APPROVED** [review]

**Pull request merged**